docs/src/references/mappers/c8y-mapper.md

```diff
@@ -1325,3 +1325,191 @@ te/device/main///cmd/log_upload/<cmd_id>
+just because a few entities required some customizations via explicit registration.
+
+The other approach would be to resort to auto-registration only when
+an explicit registration message is not delivered within a predefined window(timeout).
+But, the same timing uncertainty issues described above, applies here as well:
+what if the explicit registration message is received after the auto-registration has already happened after the timeout?
+Crashing itself isn't gonna help here either, as the entity is already auto-registered.
+
+### Final Solution
+
+The ordering problems when the mapper is live and when it is restarted would be handled differently.
+
+#### Phase 1
+
+**Live mapper**
+
+When the mapper is live, if telemetry messages for a new entity is received before its registration message,
+they are simply dropped until that registration message is processed, with the assumption that,
+telemetry data updates will repeat eventually and hence will make up for the lost ones.
+But, all non-telemetry data like twin updates, capability messages etc are cached in-memory,
+and will be processed and registered in the entity store once the registration message eventually arrives.
+Keeping them cached in-memory is not a big deal, as they are eventually going to end-up in the in-memory entity store anyway.
+
+**On mapper restart**
+
+Sticking to the same solution of dropping telemetry messages on a mapper restart might result in loosing too much data,
+when the volume of queued telemetry messages are very high on the broker,
+increasing the probability of a lot of those telemetry messages getting delivered to the mapper first,
+before their respective registration messages are.
+
+To avoid so much data loss, on restart the mapper will selectively subscribe to the registration topics first,
+process all the retained registration messages (with a timeout), after which the regular message processing will start.
+The timeout is required because MQTT offers no clear boundary/indication when all queued retained messages are delivered,
+and hence it has to be controlled with a deterministic waiting period.
+Since this timed waiting is during the startup, the impact won't be that visible to external users.
+
+**Auto registration**
+
+To avoid any conflicts between auto-registration and explicit registration,
+resulting in nested child devices from wrongly getting auto-registered as immediate child devices,
+we can just instruct the users to turn-off auto-registration if they have nested child devices in their deployment.
+Explicit registration must be used even for immediate child devices in such deployments.
+Or to say it differently, use auto-registration only in simple deployments where there are only immediate child devices.
+
+#### Phase 2
+
+To mitigate the telemetry message drop limitation of the above solution,
+cloud-specific MQTT topics can be used as secondary caches instead, for those "early messages".
```

Comment:
I'm afraid that this will only push the issue elsewhere. How using mapper private secondary caches would help?  This is only a different kind of cache (in-memory vs files vs delegated to the MQTT broker). The true issues are not specifically addressed and only *mitigated*: message re-ordering, user failing to send registration messages, ever-growing set of pending messages, bad interactions with auto-registration.

I other words, I would only consider phase 1 or a user-defined topic schema.

Reply:
> The true issues are not specifically addressed and only mitigated: message re-ordering, user failing to send registration messages, ever-growing set of pending messages, bad interactions with auto-registration.

The advantage with dedicated mapper-local topics is that it would be easier to manage with a dynamic registration or even a dynamic MQTT client. Once the registration message is received, we can have a temporary MQTT client connecting to that topic and exhaust all the queued messages on that topic, before processing other regular messages.

If the user never sends a registration message, then yes there's the risk of all those messages getting queued on these topics, until the mapper restarts next time.